File: clients/reactive-dart/lib/src/session_lifecycle.dart
import 'dart:async';
import 'dart:convert';

/// Why an authenticated application lifecycle requested a foreground cycle.
enum SessionSyncReason { login, logout }

/// Largest integer represented exactly by every supported Dart/JS/Rust host.
const int maxPortableSessionInteger = 9007199254740991;

/// Tenant-bound identity for one authenticated client session.
final class OptoSyncSessionIdentity {
  OptoSyncSessionIdentity({
    required this.subject,
    required this.tenant,
    required this.authEpoch,
  }) {
    if (subject.trim().isEmpty || utf8.encode(subject).length > 512) {
      throw ArgumentError.value(
        subject,
        'subject',
        'must be 1 through 512 bytes',
      );
    }
    if (tenant.trim().isEmpty || utf8.encode(tenant).length > 512) {
      throw ArgumentError.value(
        tenant,
        'tenant',
        'must be 1 through 512 bytes',
      );
    }
    if (authEpoch < 0 || authEpoch > maxPortableSessionInteger) {
      throw ArgumentError.value(
        authEpoch,
        'authEpoch',
        'must be from 0 through $maxPortableSessionInteger',
      );
    }
  }

  final String subject;
  final String tenant;
  final int authEpoch;

  bool sameSession(OptoSyncSessionIdentity other) =>
      subject == other.subject &&
      tenant == other.tenant &&
      authEpoch == other.authEpoch;
}

/// Durable accounting receipt returned by the application's ordinary sync loop.
///
/// A logout is drained only when every mutation present at the start of the
/// bounded cycle has an exact server acknowledgement, the local queue has
/// committed that acknowledgement, and no pending rows remain. Merely sending
/// a request, receiving a transport ACK, or timing out is never sufficient.
final class DurableSyncReceipt {
  DurableSyncReceipt({
    required this.pendingBefore,
    required this.acknowledged,
    required this.admittedDuringDrain,
    required this.pendingAfter,
    required this.checkpointCommitted,
    required this.admissionFenced,
  }) {
    if (pendingBefore < 0 ||
        acknowledged < 0 ||
        admittedDuringDrain < 0 ||
        pendingAfter < 0 ||
        pendingBefore > maxPortableSessionInteger ||
        acknowledged > maxPortableSessionInteger ||
        admittedDuringDrain > maxPortableSessionInteger ||
        pendingAfter > maxPortableSessionInteger) {
      throw ArgumentError(
        'durable sync counters must be portable integers from 0 through '
        '$maxPortableSessionInteger',
      );
    }
    if (acknowledged > pendingBefore ||
        pendingAfter != pendingBefore - acknowledged + admittedDuringDrain) {
      throw ArgumentError(
        'durable sync receipt violates queue conservation: '
        'pendingAfter must equal pendingBefore - acknowledged + '
        'admittedDuringDrain',
      );
    }
  }

  final int pendingBefore;
  final int acknowledged;

  /// Writes observed after the start snapshot. A correct logout fences these.
  final int admittedDuringDrain;
  final int pendingAfter;
  final bool checkpointCommitted;

  /// True only after the host has stopped session-scoped mutation admission.
  final bool admissionFenced;

  bool get durablyDrained =>
      pendingAfter == 0 &&
      acknowledged == pendingBefore &&
      admittedDuringDrain == 0 &&
      checkpointCommitted &&
      admissionFenced;
}

typedef AuthenticatedSync = Future<DurableSyncReceipt> Function(
  SessionSyncReason reason,
);
typedef TelemetryForceFlush = Future<void> Function();
typedef ClearSessionCredentials = Future<void> Function(
  OptoSyncSessionIdentity? session,
);

final class SessionLoginReport {
  const SessionLoginReport({
    required this.session,
    required this.syncTriggered,
    this.receipt,
    this.syncError,
    this.syncStackTrace,
  });

  final OptoSyncSessionIdentity session;
  final bool syncTriggered;
  final DurableSyncReceipt? receipt;
  final Object? syncError;
  final StackTrace? syncStackTrace;

  bool get syncSucceeded => syncTriggered && syncError == null;
}

final class SessionLogoutReport {
  const SessionLogoutReport({
    required this.hadSession,
    required this.credentialsCleared,
    this.receipt,
    this.syncError,
    this.telemetryError,
    this.credentialError,
  });

  final bool hadSession;
  final bool credentialsCleared;
  final DurableSyncReceipt? receipt;
  final Object? syncError;
  final Object? telemetryError;
  final Object? credentialError;

  bool get dataDurablyDrained =>
      !hadSession || (syncError == null && receipt?.durablyDrained == true);
  bool get telemetryFlushed => telemetryError == null;
  bool get complete =>
      dataDurablyDrained && telemetryFlushed && credentialsCleared;
}

/// Serializes authenticated login/logout boundaries for Flutter and Dart hosts.
///
/// The application still owns authentication, secure storage, the concrete
/// opto-sync loop, and the ORES OTEL providers. This coordinator only enforces
/// ordering and honest durable acknowledgement semantics.
final class AuthenticatedSessionLifecycle {
  AuthenticatedSessionLifecycle({
    required AuthenticatedSync sync,
    required TelemetryForceFlush forceFlushTelemetry,
    required ClearSessionCredentials clearCredentials,
  }) : _sync = sync,
       _forceFlushTelemetry = forceFlushTelemetry,
       _clearCredentials = clearCredentials;

  final AuthenticatedSync _sync;
  final TelemetryForceFlush _forceFlushTelemetry;
  final ClearSessionCredentials _clearCredentials;
  OptoSyncSessionIdentity? _session;
  Future<void> _tail = Future<void>.value();

  OptoSyncSessionIdentity? get session => _session;
  bool get isAuthenticated => _session != null;

  Future<T> _serialize<T>(Future<T> Function() operation) async {
    final predecessor = _tail;
    final release = Completer<void>();
    _tail = release.future;
    await predecessor;
    try {
      return await operation();
    } finally {
      release.complete();
    }
  }

  /// Records a successful authentication and immediately wakes foreground sync.
  /// Duplicate notifications for the exact same auth epoch are coalesced.
  Future<SessionLoginReport> onLogin(
    OptoSyncSessionIdentity next,
  ) => _serialize(() async {
    final current = _session;
    if (current != null && !current.sameSession(next)) {
      throw StateError(
        'logout must complete before a tenant, subject, or auth-epoch switch',
      );
    }
    if (current != null) {
      return SessionLoginReport(session: current, syncTriggered: false);
    }

    _session = next;
    try {
      final receipt = await _sync(SessionSyncReason.login);
      return SessionLoginReport(
        session: next,
        syncTriggered: true,
        receipt: receipt,
      );
    } catch (error, stackTrace) {
      // Authentication succeeded even if the first foreground cycle did
      // not. Durable queue state remains authoritative for the next wake.
      return SessionLoginReport(
        session: next,
        syncTriggered: true,
        syncError: error,
        syncStackTrace: stackTrace,
      );
    }
  });

  /// Drains session data, force-flushes app-owned telemetry, then clears auth.
  ///
  /// Every stage is attempted. Credentials are cleared even when data or
  /// telemetry delivery fails; unsafely unacknowledged data stays durable for
  /// a later authenticated recovery instead of being marked delivered.
  Future<SessionLogoutReport> onLogout() => _serialize(() async {
    final current = _session;
    DurableSyncReceipt? receipt;
    Object? syncError;
    Object? telemetryError;
    Object? credentialError;
    var credentialsCleared = false;

    if (current != null) {
      try {
        receipt = await _sync(SessionSyncReason.logout);
      } catch (error) {
        syncError = error;
      }
    }
    try {
      await _forceFlushTelemetry();
    } catch (error) {
      telemetryError = error;
    }
    try {
      await _clearCredentials(current);
      credentialsCleared = true;
    } catch (error) {
      credentialError = error;
    } finally {
      // Never retain an authenticated in-memory session after logout was
      // requested, even when platform secure-storage cleanup reports failure.
      _session = null;
    }

    return SessionLogoutReport(
      hadSession: current != null,
      credentialsCleared: credentialsCleared,
      receipt: receipt,
      syncError: syncError,
      telemetryError: telemetryError,
      credentialError: credentialError,
    );
  });
}

File: clients/desktop-rust/src/session_lifecycle.rs
//! Authenticated login/logout ordering shared with the Dart/Flutter adapter.
//!
//! Authentication and telemetry providers remain application-owned. This
//! module only makes the sync/flush/credential-clear ordering explicit and
//! prevents a transport acknowledgement from masquerading as durable delivery.

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum SessionSyncReason {
    Login,
    Logout,
}

/// Largest integer represented exactly by every supported Dart/JS/Rust host.
pub const MAX_PORTABLE_SESSION_INTEGER: u64 = 9_007_199_254_740_991;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SessionIdentity {
    pub subject: String,
    pub tenant: String,
    pub auth_epoch: u64,
}

impl SessionIdentity {
    pub fn new(
        subject: impl Into<String>,
        tenant: impl Into<String>,
        auth_epoch: u64,
    ) -> Result<Self, &'static str> {
        let subject = subject.into();
        let tenant = tenant.into();
        if subject.trim().is_empty() || subject.len() > 512 {
            return Err("subject must be 1 through 512 bytes");
        }
        if tenant.trim().is_empty() || tenant.len() > 512 {
            return Err("tenant must be 1 through 512 bytes");
        }
        if auth_epoch > MAX_PORTABLE_SESSION_INTEGER {
            return Err("auth_epoch exceeds the portable integer domain");
        }
        Ok(Self {
            subject,
            tenant,
            auth_epoch,
        })
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct DurableSyncReceipt {
    pub pending_before: u64,
    pub acknowledged: u64,
    pub admitted_during_drain: u64,
    pub pending_after: u64,
    pub checkpoint_committed: bool,
    pub admission_fenced: bool,
}

impl DurableSyncReceipt {
    pub fn new(
        pending_before: u64,
        acknowledged: u64,
        admitted_during_drain: u64,
        pending_after: u64,
        checkpoint_committed: bool,
        admission_fenced: bool,
    ) -> Result<Self, &'static str> {
        if [
            pending_before,
            acknowledged,
            admitted_during_drain,
            pending_after,
        ]
        .into_iter()
        .any(|value| value > MAX_PORTABLE_SESSION_INTEGER)
        {
            return Err("durable sync counters exceed the portable integer domain");
        }
        let expected_pending = pending_before
            .checked_sub(acknowledged)
            .and_then(|value| value.checked_add(admitted_during_drain));
        if expected_pending != Some(pending_after) {
            return Err(
                "durable sync receipt violates queue conservation: pending_after must equal pending_before - acknowledged + admitted_during_drain",
            );
        }
        Ok(Self {
            pending_before,
            acknowledged,
            admitted_during_drain,
            pending_after,
            checkpoint_committed,
            admission_fenced,
        })
    }

    #[must_use]
    pub fn durably_drained(self) -> bool {
        self.pending_after == 0
            && self.acknowledged == self.pending_before
            && self.admitted_during_drain == 0
            && self.checkpoint_committed
            && self.admission_fenced
    }
}

#[derive(Debug)]
pub struct SessionLoginReport<SyncError> {
    pub session: SessionIdentity,
    pub sync_triggered: bool,
    pub receipt: Option<DurableSyncReceipt>,
    pub sync_error: Option<SyncError>,
}

impl<SyncError> SessionLoginReport<SyncError> {
    #[must_use]
    pub fn sync_succeeded(&self) -> bool {
        self.sync_triggered && self.sync_error.is_none()
    }
}

#[derive(Debug)]
pub struct SessionLogoutReport<SyncError, TelemetryError, CredentialError> {
    pub had_session: bool,
    pub credentials_cleared: bool,
    pub receipt: Option<DurableSyncReceipt>,
    pub sync_error: Option<SyncError>,
    pub telemetry_error: Option<TelemetryError>,
    pub credential_error: Option<CredentialError>,
}

impl<SyncError, TelemetryError, CredentialError>
    SessionLogoutReport<SyncError, TelemetryError, CredentialError>
{
    #[must_use]
    pub fn data_durably_drained(&self) -> bool {
        !self.had_session
            || (self.sync_error.is_none()
                && self
                    .receipt
                    .is_some_and(DurableSyncReceipt::durably_drained))
    }

    #[must_use]
    pub fn telemetry_flushed(&self) -> bool {
        self.telemetry_error.is_none()
    }

    #[must_use]
    pub fn complete(&self) -> bool {
        self.data_durably_drained() && self.telemetry_flushed() && self.credentials_cleared
    }
}

#[derive(Debug, Default)]
pub struct AuthenticatedSessionLifecycle {
    session: Option<SessionIdentity>,
}

impl AuthenticatedSessionLifecycle {
    #[must_use]
    pub fn session(&self) -> Option<&SessionIdentity> {
        self.session.as_ref()
    }

    pub fn on_login<SyncError, Sync>(
        &mut self,
        next: SessionIdentity,
        sync: Sync,
    ) -> Result<SessionLoginReport<SyncError>, &'static str>
    where
        Sync: FnOnce(SessionSyncReason) -> Result<DurableSyncReceipt, SyncError>,
    {
        if let Some(current) = self.session.as_ref() {
            if current != &next {
                return Err("logout must complete before a tenant, subject, or auth-epoch switch");
            }
            return Ok(SessionLoginReport {
                session: current.clone(),
                sync_triggered: false,
                receipt: None,
                sync_error: None,
            });
        }

        self.session = Some(next.clone());
        let (receipt, sync_error) = match sync(SessionSyncReason::Login) {
            Ok(receipt) => (Some(receipt), None),
            Err(error) => (None, Some(error)),
        };
        Ok(SessionLoginReport {
            session: next,
            sync_triggered: true,
            receipt,
            sync_error,
        })
    }

    pub fn on_logout<SyncError, TelemetryError, CredentialError, Sync, Flush, Clear>(
        &mut self,
        sync: Sync,
        force_flush_telemetry: Flush,
        clear_credentials: Clear,
    ) -> SessionLogoutReport<SyncError, TelemetryError, CredentialError>
    where
        Sync: FnOnce(SessionSyncReason) -> Result<DurableSyncReceipt, SyncError>,
        Flush: FnOnce() -> Result<(), TelemetryError>,
        Clear: FnOnce(Option<&SessionIdentity>) -> Result<(), CredentialError>,
    {
        let current = self.session.take();
        let (receipt, sync_error) = if current.is_some() {
            match sync(SessionSyncReason::Logout) {
                Ok(receipt) => (Some(receipt), None),
                Err(error) => (None, Some(error)),
            }
        } else {
            (None, None)
        };
        let telemetry_error = force_flush_telemetry().err();
        let credential_result = clear_credentials(current.as_ref());
        let credentials_cleared = credential_result.is_ok();

        SessionLogoutReport {
            had_session: current.is_some(),
            credentials_cleared,
            receipt,
            sync_error,
            telemetry_error,
            credential_error: credential_result.err(),
        }
    }

    /// Async-host adapter with the same ordering and ownership semantics.
    pub async fn on_login_async<SyncError, Sync, SyncFuture>(
        &mut self,
        next: SessionIdentity,
        sync: Sync,
    ) -> Result<SessionLoginReport<SyncError>, &'static str>
    where
        Sync: FnOnce(SessionSyncReason) -> SyncFuture,
        SyncFuture: std::future::Future<Output = Result<DurableSyncReceipt, SyncError>>,
    {
        if let Some(current) = self.session.as_ref() {
            if current != &next {
                return Err("logout must complete before a tenant, subject, or auth-epoch switch");
            }
            return Ok(SessionLoginReport {
                session: current.clone(),
                sync_triggered: false,
                receipt: None,
                sync_error: None,
            });
        }

        self.session = Some(next.clone());
        let (receipt, sync_error) = match sync(SessionSyncReason::Login).await {
            Ok(receipt) => (Some(receipt), None),
            Err(error) => (None, Some(error)),
        };
        Ok(SessionLoginReport {
            session: next,
            sync_triggered: true,
            receipt,
            sync_error,
        })
    }

    /// Async-host logout: durable sync, provider force-flush, credential clear.
    pub async fn on_logout_async<
        SyncError,
        TelemetryError,
        CredentialError,
        Sync,
        SyncFuture,
        Flush,
        FlushFuture,
        Clear,
        ClearFuture,
    >(
        &mut self,
        sync: Sync,
        force_flush_telemetry: Flush,
        clear_credentials: Clear,
    ) -> SessionLogoutReport<SyncError, TelemetryError, CredentialError>
    where
        Sync: FnOnce(SessionSyncReason) -> SyncFuture,
        SyncFuture: std::future::Future<Output = Result<DurableSyncReceipt, SyncError>>,
        Flush: FnOnce() -> FlushFuture,
        FlushFuture: std::future::Future<Output = Result<(), TelemetryError>>,
        Clear: FnOnce(Option<SessionIdentity>) -> ClearFuture,
        ClearFuture: std::future::Future<Output = Result<(), CredentialError>>,
    {
        let current = self.session.take();
        let (receipt, sync_error) = if current.is_some() {
            match sync(SessionSyncReason::Logout).await {
                Ok(receipt) => (Some(receipt), None),
                Err(error) => (None, Some(error)),
            }
        } else {
            (None, None)
        };
        let telemetry_error = force_flush_telemetry().await.err();
        let credential_result = clear_credentials(current.clone()).await;
        let credentials_cleared = credential_result.is_ok();

        SessionLogoutReport {
            had_session: current.is_some(),
            credentials_cleared,
            receipt,
            sync_error,
            telemetry_error,
            credential_error: credential_result.err(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn identity() -> SessionIdentity {
        SessionIdentity::new("subject-1", "tenant-a", 7).expect("valid identity")
    }

    fn drained(count: u64) -> DurableSyncReceipt {
        DurableSyncReceipt::new(count, count, 0, 0, true, true).expect("valid receipt")
    }

    #[test]
    fn login_syncs_once_and_rejects_an_implicit_tenant_switch() {
        let mut lifecycle = AuthenticatedSessionLifecycle::default();
        let first = lifecycle
            .on_login(identity(), |reason| {
                assert_eq!(reason, SessionSyncReason::Login);
                Ok::<_, ()>(drained(2))
            })
            .expect("login accepted");
        assert!(first.sync_succeeded());

        let duplicate = lifecycle
            .on_login(identity(), |_| -> Result<DurableSyncReceipt, ()> {
                panic!("duplicate login must coalesce")
            })
            .expect("duplicate accepted");
        assert!(!duplicate.sync_triggered);

        assert!(lifecycle
            .on_login(
                SessionIdentity::new("subject-1", "tenant-b", 8).expect("identity"),
                |_| Ok::<_, ()>(drained(0)),
            )
            .is_err());
    }

    #[test]
    fn logout_orders_durable_sync_then_force_flush_then_credential_clear() {
        let mut lifecycle = AuthenticatedSessionLifecycle::default();
        lifecycle
            .on_login(identity(), |_| Ok::<_, ()>(drained(0)))
            .expect("login accepted");
        let order = std::cell::RefCell::new(Vec::new());
        let report = lifecycle.on_logout(
            |_| {
                order.borrow_mut().push("logout-sync");
                Ok::<_, ()>(drained(3))
            },
            || {
                order.borrow_mut().push("telemetry-force-flush");
                Ok::<_, ()>(())
            },
            |session| {
                assert_eq!(session, Some(&identity()));
                order.borrow_mut().push("credentials-clear");
                Ok::<_, ()>(())
            },
        );
        assert_eq!(
            *order.borrow(),
            ["logout-sync", "telemetry-force-flush", "credentials-clear"]
        );
        assert!(report.complete());
        assert!(lifecycle.session().is_none());
    }

    #[test]
    fn incomplete_acknowledgement_is_never_reported_as_a_drained_logout() {
        let mut lifecycle = AuthenticatedSessionLifecycle::default();
        lifecycle
            .on_login(identity(), |_| Ok::<_, ()>(drained(0)))
            .expect("login accepted");
        let report = lifecycle.on_logout(
            |_| Ok::<_, ()>(DurableSyncReceipt::new(4, 3, 0, 1, true, true).expect("receipt")),
            || Ok::<_, ()>(()),
            |_| Ok::<_, ()>(()),
        );
        assert!(!report.data_durably_drained());
        assert!(!report.complete());
        assert!(lifecycle.session().is_none());
    }

    #[test]
    fn concurrent_admission_is_accounted_but_never_called_drained() {
        let receipt = DurableSyncReceipt::new(4, 4, 1, 1, true, false).expect("receipt");
        assert!(!receipt.durably_drained());
    }

    #[test]
    fn receipt_rejects_portable_domain_overflow_without_saturation() {
        assert!(DurableSyncReceipt::new(
            MAX_PORTABLE_SESSION_INTEGER,
            0,
            1,
            MAX_PORTABLE_SESSION_INTEGER,
            true,
            false,
        )
        .is_err());
    }
}
